### PR TITLE
add implementation notes

### PR DESCRIPTION
**Pull request by mtbc**
see https://github.com/mtbc/omero-downloader/blob/overview/IMPLEMENTATION.md

### CODE REVIEW

File: IMPLEMENTATION.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+   `OptionParser` which encodes the client application's command-line
+   options. That class uses `OptionSet` as a helper for options whose
+   value represents a subset of possible choices. An instance of
+   `OptionParser.Chosen` bears the options parsed from the command line.
+
+1. OMERO.blitz's `FindChildren` to determine which OMERO model objects
+   are targeted. The `RequestManager` is a helper class for submitting
+   requests to OMERO and ensuring that the response is of the expected
+   type. A `FileMapper` instance stores the outcome of queries to
+   understand how filesets on the server relates to their images and
```
Comment (manics) [inline IMPLEMENTATION.md]: `filesets on the server relate` or `a fileset on the server relates`

File: IMPLEMENTATION.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+   pixels instances.
+
+1. `Download.downloadFiles` to implement the `-f` option's `binary` and
+   `companion`. This is where `FileManager` is used: for downloading
+   files from a raw file store provided by the server. (The class used
+   to have more methods.) It uses `LinkMakerPaths` for linking to the
```
Comment (manics) [inline IMPLEMENTATION.md:25]: `(The class used to have more methods.)` Is this relevant to someone reading the docs today?

File: IMPLEMENTATION.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+   to note their relationships. `LocalPixels` uses a couple of helpers
+   from the package `org.openmicroscopy.client.downloader.util`:
+   `TileIterator` covers all the tiles of an image and `FileIO` reads
+   and writes integers and byte arrays. Tiles are recorded in an
+   intermediate file before Bio-Formats performs the export.
+
+1. `Download.writeXmlObjects` to implement the `-f` option's
+   `ome-xml-parts` with the help of `XmlGenerator`. This also makes some
+   use of `LinkMakerPaths` for linking objects from their containers.
+   `MetadataBase` provides helpers for the other classes of the
```
Comment (manics) [inline IMPLEMENTATION.md]: Since this introduces a new package perhaps make it clearer, e.g. `metadata.MetadataBase`, or rearrange the sentence to mention the package name first?

**Review by manics — COMMENTED** [review]
A few minor points, but otherwise looks fine for a first version of the docs.

However it feels like this would be much nicer if there were links to the corresponding classes and methods, for instance by generating and linking to javadocs, or if you want to use an external docs hosting service it could be worth investigating https://javasphinx.readthedocs.io/en/latest/ to publish on readthedocs?

File: IMPLEMENTATION.md (hunk 2 above)
Comment (mtbc) [reply IMPLEMENTATION.md:25]: I thought they might wonder if there's a reason why a single modest method gets a class all of its own.

**Comment by pwalczysko** [comment]
``The code's focus on top-level OME Model units is too pervasive. For example, one may need to export annotations that are on channels.``

and

``The input and output of metadata has incomplete coverage. For example, try importing or exporting OME-XML datasets using OMERO code.``

I think it is quite important to capture annotations which are directly on Projects, Datasets and Screen/Plates/Wells. Is it captured by this strategy ? Maybe this point could be more explicit

**Comment by mtbc** [comment]
Yes, annotations on datasets can be exported with the current code.

**Comment by mtbc** [comment]
Moving this markdown file to be a Javadoc overview page would make links much easier for even offline use: created https://trello.com/c/N7Y3Dajl/62-overview-page-for-implementation accordingly.

**Review by manics — APPROVED** [review]
👍
